core/src/main/scala/kafka/zk/ZkData.scala

```diff
@@ -446,75 +448,200 @@ object StateChangeHandlers {
-    case _ => throw new IllegalArgumentException("Unknown name type:" + nameType)
+sealed trait ZkAclStore {
+  val patternType: ResourceNameType
+  val aclPath: String
+
+  def path(resourceType: ResourceType): String = s"$aclPath/$resourceType"
+
+  def path(resourceType: ResourceType, resourceName: String): String = s"$aclPath/$resourceType/$resourceName"
+
+  def changeStore: ZkAclChangeStore
+}
+
+object ZkAclStore {
+  private val storesByType = ResourceNameType.values
+    .filter(nameType => nameType != ResourceNameType.ANY && nameType != ResourceNameType.UNKNOWN)
+    .map(nameType => (nameType, create(nameType)))
+    .toMap
+
+  val stores: Iterable[ZkAclStore] = storesByType.values
+
+  val securePaths: Iterable[String] = stores
+    .flatMap(store => Set(store.aclPath, store.changeStore.aclChangePath))
+
+  def apply(patternType: ResourceNameType): ZkAclStore = {
+    storesByType.get(patternType) match {
+      case Some(store) => store
+      case None => throw new KafkaException(s"Invalid pattern type: $patternType")
+    }
   }
 
-  val aclChangePath: String = nameType match {
-    case ResourceNameType.LITERAL => "/kafka-acl-changes"
-    case ResourceNameType.PREFIXED => "/kafka-prefixed-acl-changes"
-    case _ => throw new IllegalArgumentException("Unknown name type:" + nameType)
+  private def create(patternType: ResourceNameType) = {
+    patternType match {
+      case ResourceNameType.LITERAL => LiteralAclStore
+      case _ => new ExtendedAclStore(patternType)
+    }
   }
+}
 
-  def path(resourceType: ResourceType) = s"$aclPath/$resourceType"
+object LiteralAclStore extends ZkAclStore {
+  val patternType: ResourceNameType = ResourceNameType.LITERAL
+  val aclPath: String = "/kafka-acl"
 
-  def path(resourceType: ResourceType, resourceName: String): String = s"$aclPath/$resourceType/$resourceName"
+  def changeStore: ZkAclChangeStore = LiteralAclChangeStore
+}
 
-  def changeSequenceZNode: AclChangeNotificationSequenceZNode = AclChangeNotificationSequenceZNode(this)
+class ExtendedAclStore(val patternType: ResourceNameType) extends ZkAclStore {
+  if (patternType == ResourceNameType.LITERAL)
+    throw new IllegalArgumentException("Literal pattern types are not supported")
 
-  def decode(notificationMessage: Array[Byte]): Resource = AclChangeNotificationSequenceZNode.decode(nameType, notificationMessage)
+  val aclPath: String = s"/kafka-acl-extended/${patternType.name.toLowerCase}"
+
+  def changeStore: ZkAclChangeStore = ExtendedAclChangeStore
 }
 
-object ZkAclStore {
-  val stores: Seq[ZkAclStore] = ResourceNameType.values
-    .filter(nameType => nameType != ResourceNameType.ANY && nameType != ResourceNameType.UNKNOWN)
-    .map(nameType => ZkAclStore(nameType))
+trait AclChangeNotificationHandler {
+  def processNotification(resource: Resource): Unit
+}
 
-  val securePaths: Seq[String] = stores
-    .flatMap(store => List(store.aclPath, store.aclChangePath))
+trait AclChangeSubscription extends AutoCloseable {
```

Comment:
Do we need to extend from AutoCloseable since close() doesn't throw exception?

Reply:
Yep.